lib/kv/manager/src/layout.rs

```diff
@@ -0,0 +1,142 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Block layout management.
+//!
+//! This module provides traits and implementations for managing how blocks
+//! are arranged in storage, including both contiguous and non-contiguous layouts.
+
+pub mod contiguous;
+
+use crate::dtype::DType;
+
+use derive_builder::Builder;
+use thiserror::Error;
+use tracing::instrument;
+use validator::Validate;
+
+/// Errors that can occur during layout operations
+#[derive(Debug, Error)]
+#[allow(missing_docs)]
+pub enum LayoutError {
+    #[error("Invalid configuration: {0}")]
+    InvalidConfig(String),
+
+    #[error("Validation failed: {0}")]
+    ValidationError(#[from] validator::ValidationErrors),
+
+    #[error("Invalid block index: {0}")]
+    InvalidBlockIndex(usize),
+
+    #[error("Invalid layer index: {0}")]
+    InvalidLayerIndex(usize),
+
+    #[error("Operation failed: {0}")]
+    OperationFailed(String),
+}
+
+/// Storage pattern for layers
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum LayerConfiguration {
+    /// All layers are contiguous in memory [n_layers, ...]
+    FullyContiguous,
+
+    /// Each layer is stored separately with a common stride between blocks
+    /// in different layers
+    LayerContiguousWithCommonStride,
+
+    /// Each layer is stored separately with no guaranteed stride
+    LayerContiguousWithSeparateStride,
+
+    /// Each page is stored separately with no guaranteed stride
+    PageContiguousWithSeparateStride,
+
+    /// NullLayout
+    /// Used for testing and debugging
+    Null,
+}
+
+/// Core trait for block layouts
+pub trait BlockLayout: Send + Sync + std::fmt::Debug {
+    /// Returns the total number of blocks this layout manages
+    fn num_blocks(&self) -> usize;
+
+    /// Returns the number of layers per block
+    fn num_layers(&self) -> usize;
+
+    /// Returns the size of each block in bytes
+    fn page_size(&self) -> usize;
+
+    /// Returns the inner dimension size
+    fn inner_dim(&self) -> usize;
+
+    /// Get the memory region for a specific page [page_size, inner_dim]
+    fn get_memory_region(&self, block_idx: usize, layer_idx: usize) -> Result<u64, LayoutError>;
```

Comment:
How does the `u64` in the response become a memory region? Do we convert it to a pointer? Is it an ID? Something else?

Reply:
Good point. In the NIXL crate we have a MemoryRegion. 

I can refactor this to the memory region object once the NIXL crate is published. 

But yes, we should put some kind of string typing on that u64. 

Let's resolve it in a follow up issue if that is OK.

Reply:
Sound good to me.